File: 백준/1000/1051/1051.py
# https://www.acmicpc.net/problem/1051

# 3 5
# 42101
# 22100
# 22101

import sys
input = sys.stdin.readline
n, m = map(int, input().strip().split())
arr = [input().strip() for _ in range(n)]
# print(n, m)
size = 1
flag = False
for i in range(min(n, m), 1, -1):
    # print(i)
    for j in range(n-i+1):
        for k in range(m-i+1):
            # print(j, k, i)
            if arr[j][k] == arr[j][k+i-1] == arr[j+i-1][k] == arr[j+i-1][k+i-1]:
                size = i
                flag = True
                break

        if flag:
            break
    if flag:
        break

print(size**2)

File: 백준/1000/1051/1051.js
// https://www.acmicpc.net/problem/1051

const fs = require('fs');
const stdin = (process.platform === 'linux'
    ? fs.readFileSync('/dev/stdin').toString()
    : `3 5
    42101
    22100
    22101`
).split('\n');

const input = (() => {
    let line = 0;
    return () => stdin[line++];
})();

const [N, M] = input().split(' ').map(v => +v);
const arr = [];
for (let i = 0; i < N; i++) {
    arr.push(input().trim().split('').map(v => +v));
}

const sizeCheck = (x, y, size) => {
    const flag = arr[x][y];
    const result = (flag === arr[x + size - 1][y] && flag === arr[x][y + size - 1] && flag === arr[x + size - 1][y + size - 1]);
    return result;
}

let answer = 1;
for (let squareSize = 2; squareSize <= Math.min(N, M); squareSize++) {
    for (let j = 0; j <= (M - squareSize); j++) {
        for (let k = 0; k <= (N - squareSize); k++) {
            console.log(squareSize, k, j)
            // console.log(check(k, j, squareSize))
            if (sizeCheck(k, j, squareSize)) {
                console.log(k, j, squareSize, check(k, j, squareSize))
                answer = Math.max(answer, squareSize ** 2);
                break;
            }
        }
    }
    console.log()
}
console.log(answer);